src/transport.c

```diff
@@ -176,21 +176,72 @@ fullpacket(LIBSSH2_SESSION * session, int encrypted /* 1 or 0 */ )
         if(encrypted) {
 
             /* Calculate MAC hash */
-            session->remote.mac->hash(session, macbuf,  /* store hash here */
-                                      session->remote.seqno,
-                                      p->init, 5,
-                                      p->payload,
-                                      session->fullpacket_payload_len,
-                                      &session->remote.mac_abstract);
+            int etm = session->remote.mac->etm;
```

Comment:
My attempt at etm support was considerably less change in fullpacket(). Here's the changes I've made for comparison:

`int compressed;
    int required_full_packet = encrypted && session->remote.mac->etm == 1;
	const LIBSSH2_MAC_METHOD *remote_mac = NULL;
	
	if ( required_full_packet ) {
		if ( session->remote.crypt->auth_len == 0 )
			remote_mac = session->remote.mac;
	} else if ( encrypted ) {
		remote_mac = session->remote.mac;
	}
	
    if (session->fullpacket_state == libssh2_NB_state_idle) {
		
		if (!required_full_packet || (required_full_packet && remote_mac == NULL) )
            session->fullpacket_macstate = LIBSSH2_MAC_CONFIRMED;
		
		session->fullpacket_payload_len = p->packet_length - 1;

        if (encrypted) {
			
            if (!required_full_packet && remote_mac) {
                /* Calculate MAC hash */
                remote_mac->hash(session, macbuf,  /* store hash here */
                              session->remote.seqno,
                              p->init, 5,
                              p->payload,
                              session->fullpacket_payload_len,
                              &session->remote.mac_abstract);

                /* Compare the calculated hash with the MAC we just read from
                 * the network. The read one is at the very end of the payload
                 * buffer. Note that 'payload_len' here is the packet_length
                 * field which includes the padding but not the MAC.
                 */
                if (memcmp(macbuf, p->payload + session->fullpacket_payload_len,
                           session->remote.mac->mac_len)) {
                    session->fullpacket_macstate = LIBSSH2_MAC_INVALID;
                }
            }
            else {
                //etm trim off padding byte from payload
                memmove(p->payload, &p->payload[1], p->packet_length - 1);
            }
        }
`

Reply:
Your change is indeed much nicer @willco007.

I merged in the latest commits from master and replaced my change with yours but it requires `auth_len` in `_LIBSSH2_CRYPT_METHOD` failing with:
````
[ 19%] Building C object src/CMakeFiles/libssh2.dir/transport.c.o
/Users/ander/opgaver/ShellFish/related/encrypt-then-mac/libssh2/src/transport.c:175:36: error: no member named 'auth_len' in 'struct _LIBSSH2_CRYPT_METHOD'
        if (session->remote.crypt->auth_len == 0) {
            ~~~~~~~~~~~~~~~~~~~~~  ^
1 error generated.
````

`_LIBSSH2_CRYPT_METHOD` is currently: 
https://github.com/libssh2/libssh2/blob/821d50dad313b53fb2782f26aec1f52f1be34fc0/src/libssh2_priv.h#L926-L948

I presume `auth_len` is crucial for ciphers where encryption and authentication are not separate but I'm unsure how to proceed with this pull request.